x-pack/plugins/observability/server/services/slo/tasks/summary_cleanup_task.ts

```diff
@@ -0,0 +1,212 @@
+  private abortController = new AbortController();
+  private logger: Logger;
+  private taskManager?: TaskManagerStartContract;
+  private soClient?: SavedObjectsClientContract;
+  private esClient?: ElasticsearchClient;
+
+  constructor(taskManager: TaskManagerSetupContract, logger: Logger) {
+    this.logger = logger;
+
+    taskManager.registerTaskDefinitions({
+      [TASK_TYPE]: {
+        title: 'SLO Definitions Cleanup Task',
+        timeout: '3m',
+        maxAttempts: 1,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              return this.runTask();
+            },
+
+            cancel: async () => {
+              this.abortController.abort('[SLO] Definitions clean up Task timed out');
+            },
+          };
+        },
+      },
+    });
+  }
+
+  runTask = async () => {
+    const runAt = new Date().toISOString();
+
+    if (this.soClient && this.esClient) {
+      const finder = this.soClient.createPointInTimeFinder<Pick<StoredSLO, 'id' | 'revision'>>({
+        type: SO_SLO_TYPE,
+        perPage: 1000,
+        fields: ['id', 'revision'],
+      });
+      let searchAfterKey: SortResults | undefined;
+      let soIdsToCheck: string[] = [];
+      let sloSummaryIdsToDelete: string[] = [];
+
+      for await (const response of finder.find()) {
+        const soItems = response.saved_objects.map(
+          (so) => `${so.attributes.id}${SEPARATOR}${so.attributes.revision}`
+        );
+
+        const { sloSummaryIds, searchAfter } = await this.fetchSloSummariesIds(searchAfterKey);
+        searchAfterKey = searchAfter;
+
+        const { missingSummaryIds, missingSOIds } = findMissingItems(
+          soItems.concat(soIdsToCheck),
+          sloSummaryIds.concat(sloSummaryIdsToDelete)
+        );
+
+        soIdsToCheck = missingSOIds;
+        sloSummaryIdsToDelete = missingSummaryIds;
+        if (searchAfterKey === undefined) {
+          break;
+        }
+      }
+
+      await finder.close();
+
+      if (sloSummaryIdsToDelete.length > 0) {
+        this.logger.info(
+          `[SLO] Deleting ${sloSummaryIdsToDelete.length} SLO Summaries from the summary index`
+        );
+        this.logger.info(JSON.stringify(sloSummaryIdsToDelete));
+
+        await this.esClient.deleteByQuery({
+          index: SLO_SUMMARY_DESTINATION_INDEX_PATTERN,
+          query: {
+            bool: {
+              should: getDeleteQueryFilter(sloSummaryIdsToDelete.sort()),
+            },
+          },
+        });
+      }
+    }
+
+    return { state: { lastRunAt: runAt } };
+  };
+
+  fetchSloSummariesIds = async (searchAfter?: SortResults) => {
+    if (this.esClient) {
+      const result = await this.esClient.search({
+        size: 2000,
+        index: SLO_SUMMARY_DESTINATION_INDEX_PATTERN,
+        fields: ['slo.id', 'slo.revision'],
+        collapse: {
+          field: 'slo.id',
+        },
```

Comment:
The collapse groups results by `slo.id` which hides any documents with a previous `revision` number that we would like to clean from the summary index.

For example, index these two documents (same slo id but different revision)
```
POST .slo-observability.summary-v3/_doc
{
  "summaryUpdatedAt": "2023-11-24T19:36:17.909929Z",
  "goodEvents": 34940,
  "errorBudgetEstimated": false,
  "totalEvents": 36397,
  "errorBudgetRemaining": 0.33282047055160324,
  "errorBudgetConsumed": 0.6671795294483968,
  "isTempDoc": false,
  "latestSliTimestamp": "2023-11-24T19:33:00.000Z",
  "service": {
    "environment": null,
    "name": null
  },
  "slo": {
    "indicator": {
      "type": "sli.kql.custom"
    },
    "timeWindow": {
      "duration": "7d",
      "type": "rolling"
    },
    "instanceId": "mail.at",
    "name": "[Copy] test by domain",
    "description": "test",
    "id": "79cfd3db-242d-4184-a696-31c221260f52",
    "budgetingMethod": "occurrences",
    "groupBy": "url.domain",
    "revision": 999,
    "tags": [
      "test",
      "prod"
    ],
    "objective": {
      "target": 0.94
    }
  },
  "errorBudgetInitial": 0.06000000000000005,
  "transaction": {
    "name": null,
    "type": null
  },
  "sliValue": 0.9599692282330962,
  "statusCode": 4,
  "status": "HEALTHY"
}


POST .slo-observability.summary-v3/_doc
{
  "summaryUpdatedAt": "2023-11-24T19:36:17.909929Z",
  "goodEvents": 34940,
  "errorBudgetEstimated": false,
  "totalEvents": 36397,
  "errorBudgetRemaining": 0.33282047055160324,
  "errorBudgetConsumed": 0.6671795294483968,
  "isTempDoc": false,
  "latestSliTimestamp": "2023-11-24T19:33:00.000Z",
  "service": {
    "environment": null,
    "name": null
  },
  "slo": {
    "indicator": {
      "type": "sli.kql.custom"
    },
    "timeWindow": {
      "duration": "7d",
      "type": "rolling"
    },
    "instanceId": "mail.at",
    "name": "[Copy] test by domain",
    "description": "test",
    "id": "79cfd3db-242d-4184-a696-31c221260f52",
    "budgetingMethod": "occurrences",
    "groupBy": "url.domain",
    "revision": 1,
    "tags": [
      "test",
      "prod"
    ],
    "objective": {
      "target": 0.94
    }
  },
  "errorBudgetInitial": 0.06000000000000005,
  "transaction": {
    "name": null,
    "type": null
  },
  "sliValue": 0.9599692282330962,
  "statusCode": 4,
  "status": "HEALTHY"
}
```

Now try to search with the collapse field:

```
POST .slo-observability.summary-v3*/_search
{
  "size": 2000,
  "fields": [
    "slo.id",
    "slo.revision"
  ],
  "collapse": {
    "field": "slo.id"
  },
  "_source": false,
  "sort": [
    {
      "slo.id": {
        "order": "desc"
      },
      "slo.revision": {
        "order": "desc"
      },
      "slo.instanceId": {
        "order": "desc"
      }
    }
  ]
}
```

You should get one result for the previous two document, for the latest revision number. But in this precise case, we want to get remove the document with the slo.revision = 1 (orphan from an update)

Reply:
you are right, collapse isn't needed here and yeah it essentially breaks the query.